dbms/src/DataStreams/ParallelParsingBlockInputStream.cpp

```diff
@@ -106,52 +101,100 @@ void ParallelParsingBlockInputStream::parserThreadFunction(size_t current_unit_n
 
     std::unique_lock lock(mutex);
-    const auto current_unit_number = reader_ticket_number % max_threads_to_use;
-    auto & unit = processing_units[current_unit_number];
-
-    reader_condvar.wait(lock, [&](){ return unit.status == READY_TO_READ || is_exception_occured || finished; });
+    if (!background_exception)
+    {
+        background_exception = std::current_exception();
+    }
+    finished = true;
+    reader_condvar.notify_all();
+    segmentator_condvar.notify_all();
+}
 
-    /// Check for an exception and rethrow it
-    if (is_exception_occured)
+Block ParallelParsingBlockInputStream::readImpl()
+{
+    if (isCancelledOrThrowIfKilled() || finished)
     {
-        lock.unlock();
-        cancel(false);
-        rethrowFirstException(exceptions);
+        /**
+          * Check for background exception and rethrow it before we return.
+          */
+        std::unique_lock lock(mutex);
+        if (background_exception)
+        {
+            lock.unlock();
+            cancel(false);
+            std::rethrow_exception(background_exception);
+        }
+
+        return Block{};
     }
 
-    res = std::move(unit.block_ext.block.at(internal_block_iter));
-    last_block_missing_values = std::move(unit.block_ext.block_missing_values[internal_block_iter]);
+    const auto current_unit_number = reader_ticket_number % processing_units.size();
+    auto & unit = processing_units[current_unit_number];
 
-    if (++internal_block_iter == unit.block_ext.block.size())
+    if (!next_block_in_current_unit.has_value())
     {
-        if (unit.is_last)
+        // We have read out all the Blocks from the current Processing Unit,
+        // time to move to the next one. Wait for it to become ready.
+        std::unique_lock lock(mutex);
+        reader_condvar.wait(lock, [&](){ return unit.status == READY_TO_READ || finished; });
+
+        if (finished)
         {
-            //In case that all data was read we don't need to cancel.
-            finished = true;
-            return res;
+            /**
+              * Check for background exception and rethrow it before we return.
+              */
+            if (background_exception)
+            {
+                lock.unlock();
+                cancel(false);
+                std::rethrow_exception(background_exception);
+            }
+
+            return Block{};
         }
 
-        internal_block_iter = 0;
+        assert(unit.status == READY_TO_READ);
+
+        if (unit.is_past_the_end)
+        {
+            // No more data.
+            return Block{};
+        }
+        next_block_in_current_unit = 0;
+    }
+
+    assert(next_block_in_current_unit);
+    assert(!unit.is_past_the_end);
+
+    Block res = std::move(unit.block_ext.block.at(*next_block_in_current_unit));
+    last_block_missing_values = std::move(unit.block_ext.block_missing_values[*next_block_in_current_unit]);
+
+    ++*next_block_in_current_unit;
```

Comment:
OMG. [Maybe this is better?](https://en.cppreference.com/w/cpp/utility/optional/value)